R/io-csv-functions.R

```diff
@@ -0,0 +1,253 @@
+#' @param decimal_comma Parse floats using a comma as the decimal separator
+#' instead of a period.
+#' @param glob Expand path given via globbing rules.
+#' @param storage_options Named vector containing options that indicate how to
+#' connect to a cloud provider. The cloud providers currently supported are
+#' AWS, GCP, and Azure.
+#' See supported keys here:
+#' * [aws](https://docs.rs/object_store/latest/object_store/aws/enum.AmazonS3ConfigKey.html)
+#' * [gcp](https://docs.rs/object_store/latest/object_store/gcp/enum.GoogleConfigKey.html)
+#' * [azure](https://docs.rs/object_store/latest/object_store/azure/enum.AzureConfigKey.html)
+#' * Hugging Face (`hf://`): Accepts an API key under the token parameter
+#'   `c(token = YOUR_TOKEN)` or by setting the `HF_TOKEN` environment
+#'   variable.
+#'
+#' If `storage_options` is not provided, Polars will try to infer the
+#' information from environment variables.
+#' @param credential_provider Provide a function that can be called to provide
+#' cloud storage credentials. The function is expected to return a dictionary
+#' of credential keys along with an optional credential expiry time.
+#' @param retries Number of retries if accessing a cloud instance fails.
+#' @param file_cache_ttl Amount of time to keep downloaded cloud files since
+#' their last access time, in seconds. Uses the `POLARS_FILE_CACHE_TTL`
+#' environment variable (which defaults to 1 hour) if not given.
+#' @param include_file_paths Include the path of the source file(s) as a column
+#' with this name.
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' my_file <- tempfile()
+#' write.csv(iris, my_file)
+#' lazy_frame <- pl$scan_csv(my_file)
+#' lazy_frame$collect()
+#' unlink(my_file)
+pl__scan_csv <- function(
+    source,
+    ...,
+    has_header = TRUE,
+    separator = ",",
+    comment_prefix = NULL,
+    quote_char = '"',
+    skip_rows = 0,
+    schema = NULL,
+    schema_overrides = NULL,
+    null_values = NULL,
+    missing_utf8_is_empty_string = FALSE,
+    ignore_errors = FALSE,
+    cache = FALSE,
+    infer_schema = TRUE,
+    infer_schema_length = 100,
+    n_rows = NULL,
+    encoding = c("utf8", "utf8-lossy"),
+    low_memory = FALSE,
+    rechunk = FALSE,
+    skip_rows_after_header = 0,
+    row_index_name = NULL,
+    row_index_offset = 0,
+    try_parse_dates = FALSE,
+    eol_char = "\n",
+    raise_if_empty = TRUE,
+    truncate_ragged_lines = FALSE,
+    decimal_comma = FALSE,
+    glob = TRUE,
+    storage_options = NULL,
+    retries = 2,
+    file_cache_ttl = NULL,
+    include_file_paths = NULL) {
+  wrap({
+    check_dots_empty0(...)
+    check_character(source)
+    if (length(source) == 0) {
+      abort("`source` must have length > 0.")
+    }
+    check_list_of_polars_dtype(schema_overrides, allow_null = TRUE)
+    check_list_of_polars_dtype(schema, allow_null = TRUE)
+    check_named_list(schema_overrides, allow_null = TRUE)
+    check_named_list(schema, allow_null = TRUE)
```

Comment:
Do you think it's worth adding a `check_has_names` param in `check_list_of_polars_dtype()` instead of having 2 functions per param?

Reply:
In my opinion, it is not a good idea to enforce names with something like a schema.
The limitation of not being able to use empty string in the name prevents things like the following from working (It works on the branch main for now):

```r
pl$DataFrame(1:5, foo = letters[1:5]) |>
  as.data.frame() |>
  arrow::write_dataset("data", format = "arrow", partitioning = "")
pl$read_ipc("data", hive_schema = list(pl$UInt32))
```